include/openssl/trace.h

```diff
@@ -261,32 +285,51 @@ void OSSL_trace_end(int category, BIO *channel);
  *                42, "What do you get when you multiply six by nine?");
  */
 
-# define OSSL_TRACEV(category, args) \
-    OSSL_TRACE_BEGIN(category) \
-        BIO_printf args; \
-    OSSL_TRACE_END(category)
+# if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 199901)
+#  define OSSL_FUNC __func__
+# elif defined(__STDC__) && defined(PEDANTIC)
+#  define OSSL_FUNC "(no function name due to PEDANTIC)"
+# elif defined(WIN32) || defined(__GNUC__) || defined(__GNUG__)
+#  define OSSL_FUNC __FUNCTION__
+# elif defined(__FUNCSIG__)
+#  define OSSL_FUNC __FUNCSIG__
+# else
+#  define OSSL_FUNC "(unknown function)"
+# endif
+# define OSSL_TRACEV(category, level, args) \
+    do { \
+        BIO *trc_out = OSSL_trace_begin(OSSL_TRACE_CATEGORY_##category, \
+                                        OSSL_TRACE_LEVEL_##level); \
+ \
+        if (trc_out != NULL) { \
+            BIO_printf(trc_out, "%s:%s:%d:%s:", OSSL_FUNC, OPENSSL_FILE, \
+                       OPENSSL_LINE, \
+                       OSSL_trace_get_verbosity_name(OSSL_TRACE_LEVEL_##level)); \
+            BIO_printf args; \
+        } \
+    OSSL_TRACE_END(category) \
 
 # define OSSL_TRACE(category, text) \
-    OSSL_TRACEV(category, (trc_out, "%s", text))
+    OSSL_TRACEV(category, UNDEFINED, (trc_out, "%s", text))
```

Comment:
The trace level MUST be part of the trace macros, too. Otherwise, the addition of trace levels does not make sense.